### DISCUSSION TITLE
Have we thought about bundling an external solution explorer?

### DISCUSSION THREAD

**Discussion by farlee2121** [ionide/ionide-vscode-fsharp#1850] [Ideas]
The solution explorer seems like a good bit of code to maintain and a source of many feature requests.
It also seems like separable sub-problem.

Has bundling a 3rd party solution explorer been considered? 
Are there constraints or benefits that make a first party explorer preferable?
- I noticed that project edits trickle down to FSAC. Not sure if that's critical.

## Possible 3rd-party Explorer
[vscode-solution-explorer](https://marketplace.visualstudio.com/items?itemName=fernandoescolar.vscode-solution-explorer) is a dedicated solution explorer project that I use it for C#. It doesn't need much lift to support F# and it already handles significant feature lifts like
- package management
- adding project references
- drag and drop
- Handling all project item types (i.e. Compile, Content, None) including linked files
- handling both new and old project file formats (not sure if Ionide supports this)

I recently contributed a few tweaks to smooth the F# experience ([showing ordered files](https://github.com/fernandoescolar/vscode-solution-explorer/pull/268), [open project file on click](https://github.com/fernandoescolar/vscode-solution-explorer/pull/269)). They haven't been deployed yet though.

I think the main lift would be move file up/down and add file above/below.

**Comment by farlee2121** [comment]
The creator of vscode-solution-explorer also [just added move file up/down for files in F# projects](https://github.com/fernandoescolar/vscode-solution-explorer/issues/270#issuecomment-1468791650).

**Comment by baronfel** [comment]
That's certainly an idea but there are a lot of open questions I have. The primary ones are:
* The explorer we have is a projection of data that we already have to acquire for other purposes (getting project information to power the compilation) - is the data from this external solution explorer consumable/sharable in the same way? Can that data be ingested into FSAC in some way?
* What should FSAC do about clients that don't have access to this VSCode-specific extension? Vim, Emacs, etc won't have this extension available and so FSAC would continue to need its own ability to unpack and understand projects, so I'm not sure that we'd even be able to remove the concept entirely from FSAC.
* extensibility - if we want to add items to project right click menus, is there a way to extend this external component easily?

**Reply by Krzysztof-Cieslak** [reply]
I've briefly scanned through the source code and it seems to me that the solution explorer doesn't perform any kind of proper MsBuild evaluation but rather than just parses project files as xml. 

As such, it doesn't seem like a good fit for us.

**Reply by baronfel** [reply]
Yes, that's a hard-requirement. Does this project handle concepts like Directory.Build.props/targets?  If not then that's a hard pass.

**Reply by farlee2121** [reply]
These are great questions!

Short version: I may have wrong assumptions. I assumed Ionide could coordinate with a standalone explorer via edits to the project file, since both watch the project file for edits. I also assumed the explorer experience wouldn't diverge much from what exists, which is almost and could be brought to parity for vscode-solution-explorer, and I also think we could extend some experiences like context menus.

Q: The explorer we have is a projection of data....
R: As you already discovered, I don't think they provide project information in a way other extensions could access.

Q: What should FSAC do about clients ...
R: I wasn't thinking we'd modify the underlying FSAC experience and didn't realize the connection between the explorer and other editor features. My assumption was that the solution explorer would be standalone and coordinate with other features through changes to the project files, since we already have to listen for direct project file edits.

Q: extensibility ...
R: The plugin does supply context (i.e. via vs code's `setContext`) that I think we could use to add our own menu items and buttons to the tree view. There is (as of this week) a way to specify features just for F# projects. This has already been used to add move file up/down. Add file above/below could also be added without much pain. I assumed (perhaps incorrectly) that the F# explorer tree view experience wouldn't diverge much from what is currently seen in Ionide or Visual Studio. vscode-solution-explorer can reach parity with both of those UI experiences and is very close. 

Q: Does this project handle concepts like Directory.Build.props/targets?
R: My understanding of MSBuild extension isn't superb. I'll try to explore this more and share my findings

**Reply by farlee2121** [reply]
I should also clarify that when I talk about bundling a third party explorer, I mean something like an [extension pack](https://code.visualstudio.com/api/references/extension-manifest#extension-packs) or an `extensionDependencies` in the [manifest](https://code.visualstudio.com/api/references/extension-manifest#fields).

The value I would hope for would be off-loading the visual solution tree view experience to a dedicated project.
This wouldn't deliver on the value if we expect coordination between the explorer and other Ionide features to need deeper integration than watching the same project file (and maybe some added context menu items or buttons)

**Reply by baronfel** [reply]
No worries at all @farlee2121 - it's a good question to ask!  IMO a reusable, embeddable project-system component would be an amazing help, but sadly one does not exist for .NET in VSCode :(

**Reply by farlee2121** [reply]
I noticed the project system history was a complicated when I tried digging into hot reload. 
It seems like a lot of tooling is built around Workspaces and MSBuildWorkspace, but F# support isn't included.
I'd be curious how Visual Studio manages it.

**Reply by farlee2121** [reply]
@baronfel I ran some simple tests and it appears vscode-solution-explorer does not consider Directory.Build.props.

**Reply by farlee2121** [reply]
I feel a bit silly. I just now discovered the [nuget package for MsBuild](https://www.nuget.org/packages/Microsoft.Build) which allows project parsing and that the [visual studio project-system](https://github.com/dotnet/project-system) is open source. 

I'm sure it's not news to the Ionide team, but figured I'd shared since I pondered about it here. Not sure why I didn't find those in my previous project system dive.

**Reply by baronfel** [reply]
Yep! We already use the former (via the proj-info library) to do project evaluation, since there is a _lot_ of hard-earned knowledge into how to do that right, and IIRC the latter isn't really architected to be a drop-in thing - it's very tied to VS APIs and may not be easily embeddable. There's also the 'sharing' part I mentioned above - ideally to foster C#<->F# interop there would be _one_ project system that each language server could be a _client_ of, and there's no project-system component in VSCode to share IIRC.

**Reply by TheAngryByrd** [reply]
@farlee2121 [this is the library](https://github.com/ionide/proj-info) @baronfel is referring to which in turn uses Microsoft.Build.

**Reply by farlee2121** [reply]
Thanks! 

If I understand right, the Ionide tree view gets it's information through FSAC which is leaning on proj-info, which in turn leans on Microsoft.Build.

**Reply by TheAngryByrd** [reply]
Yep! Here's super high level diagram on what depends on what: https://ionide.io/#toolchain

**Reply by farlee2121** [reply]
Thats a helpful diagram! I've been looking at some of the "up for grabs" issues and that should help me get oriented.

**Comment by farlee2121** [comment]
I created [an extension pack](https://marketplace.visualstudio.com/items?itemName=spencer-farley.ionide-and-vscode-solution-explorer) that demonstrates what I originally envisioned. The pack is published if you want to try it out. The F# features for vscode-solution-explorer are also released.

The pack very simply installs both Ionide and vscode-solution-explorer and hides the Ionide tree view by default (`FSharp.enableTreeView` = false). 
This way the 3rd-party explorer essentially replaces the default Ionide project explorer tree view.

If we ever wanted to officially offload the explorer tree view, we could achieve a similar result using `extensionDependencies` instead of an extension pack. As previously mentioned though, this tree view does not respect Directory.Build.props or targets. It is manipulates project xml directly.
